DOC: [META-372] Add Metadata Store logging documentation (#2)
Pivotal-Jeff-Jun/docs-tap: scst-store/logs.md

> # Configuring and Understanding Store Logs

Reviewer: ```suggestion
# Configuring and Understanding Logs
```

This topic covers how to configure the Supply Chain Security Tools - Store to output particular log information, as well on how to interpret them.

## Log Levels

There are six log levels that the Supply Chain Security Tools - Store supports. 

| Level   |            Description                      |
| ------- | ------------------------------------------- |
| Trace   | Output extended debugging logs              |
| Debug   | Output standard debugging log               |
| More    | Output more verbose informational logs      |
| Default | Output standard informational logs          |
| Less    | Outputs less verbose informational logs     |
| Minimum | Outputs a minimal set of informational logs |

When the Store is deployed at a specific log level, all logs of that level and lower will be outputted to the console. For example, setting the log level to `More` will output logs from `Minimal` to `More`, while `Debug` and `Trace` logs will be muted.

Currently, the application logs output at these levels:

> * Minimum - The Store does not output any logs

Reviewer: ```suggestion

* Minimum - does not output any logs.
```

Reviewer: Need new line before a list (docs toolchain bug)

* Less - Outputs a single log line indicating the current log level the Metadata Store is configured to when the application starts.
* Default - Outputs API endpoint access information.
* Debug - Outputs API endpoint payload information, both for requests and responses.

> Other log levels do not output any other log information. However, they are there for future extensibility.

Reviewer: ```suggestion
Other log levels do not output any additional log information. However, they are there for future extensibility.
```

If no log level is specified when the Store is installed, the log level is set to `default`.

## API Endpoint Log Output

When an API endpoint handles a request, the Store generates between two and five log lines. They are:

> 1. When the endpoint receives a request, it outputs a "Processing request" line. This log line is shown at the `default` log level

Reviewer: ```suggestion

1. When the endpoint receives a request, it outputs a "Processing request" line. This log line is shown at the `default` log level.
```

Reviewer: Need new line before a list (docs toolchain bug)

1. If the endpoint includes query or path parameters, it outputs a "Request parameters" line. This line logs the parameters being passed in the request. This line is shown at the `default` log level.
1. If the endpoint takes in a request body, it outputs a "Request body" line. This line outputs the entire request body as a string. This line is shown at the `debug` log level.

> 1. When the endpoint returns a response, it outputs a "Request response" line. This line is shown at the `default` log leve.

Reviewer 2: typo: `leve` => `level`

> 1. If the endpoint returns a response body, it outputs a second "Request response" line with an extra key `payload`, and its value is set to the full response body. This line is shown at the `debug` log level

Reviewer: ```suggestion
1. If the endpoint returns a response body, it outputs a second "Request response" line with an extra key `payload`, and its value is set to the full response body. This line is shown at the `debug` log level.
```

Reviewer: Inconsistent ending the sentence with a period or no period. Stick with one style.

The key name are the name of the query or path parameter, while the value are set to the value of those parameters, in string format.

For example, the following log line
```text
I1122 20:30:21.869791       1 images.go:34] MetadataStore "msg"="Request parameters" "endpoint"="/api/images?digest=sha256%3A20521f76ff3d27f436e03dc666cc97a511bbe71e8e8495f851d0f4bf57b0bab6" "hostname"="metadata-store-app-564f8995c8-r8d6n" "method"="GET" "digest"="sha256:20521f76ff3d27f436e03dc666cc97a511bbe71e8e8495f851d0f4bf57b0bab6" "id"=0
```
contains the `digest` and `id` key, which represents the respective `digest` and `id` query parameters, as well as their values. 

These key/value pairs shows up in all subsequent log lines of the same call. For example:
```text
I1122 20:30:21.878749       1 images.go:56] MetadataStore "msg"="Request response" "digest"="sha256:20521f76ff3d27f436e03dc666cc97a511bbe71e8e8495f851d0f4bf57b0bab6" "endpoint"="/api/images?digest=sha256%3A20521f76ff3d27f436e03dc666cc97a511bbe71e8e8495f851d0f4bf57b0bab6" "hostname"="metadata-store-app-564f8995c8-r8d6n" "id"=0 "method"="GET" "code"=200 "response"="OK"
```

This is done to ensure two things:

> 1. The application interprets the values of the query/path parameters correctly

Reviewer: ```suggestion

1. The application interprets the values of the query/path parameters correctly.
```

Reviewer: Need new line before a list (docs toolchain bug)

> 2. Help figure out which log lines are associated with a particular API request. Since there can be a number of simultaneous endpoint calls, this is a first attempt at trying to group logs by specific calls. An improved way to identify and group log by specific request is in the work for a future release.

Reviewer: ```suggestion
2. Help figure out which log lines are associated with a particular API request. Since there can be a number of simultaneous endpoint calls, this is a first attempt at trying to group logs by specific calls.
```

Reviewer: Probably don't need to mention future plans in this doc unless it's very helpful.

##### API Payload Log Output
As mentioned at the start of this section, by setting the log level to `debug`, the Store logs the body payload data for both the request and response of an API call.

> Using `debug` log level, over `default`, for this information was done for the following two reasons:

Reviewer: ```suggestion
Using `debug` log level, instead of `default`, to display this information was done for the following two reasons:
```

> * Body payloads can be very large as they contain full cycloneDX and SBOM information. By moving the payload information at this level, it helps keep the production log output to a reasonable size.

Reviewer: ```suggestion

* Body payloads can be very large as they contain full CycloneDX and SBOM information. By moving the payload information at this level, it helps keep the production log output to a reasonable size.
```

Reviewer: Need new line before a list (docs toolchain bug)

> * Some of this contained in these payloads may be sensitive, and may not want to be exposed in production environment logs.

Reviewer: ```suggestion
* Some information contained in these payloads may be sensitive, and the user may not want them to be exposed in production environment logs.
```

> This decision may be revised in future releases.

Reviewer: ```suggestion
```

Reviewer: No need to mention future plans unless it's really helpful.
